src/services/geometry/richgeo/ReadoutGeo.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright 2023, Christopher Dilks, Luigi Dello Stritto
+// Subject to the terms in the LICENSE file found in the top-level directory.
+//
+//
+
+#include "ReadoutGeo.h"
+
+// constructor
+richgeo::ReadoutGeo::ReadoutGeo(std::string detName_, dd4hep::Detector *det_, bool verbose_)
+  : m_detName(detName_), m_det(det_), m_log(Logger::Instance(verbose_))
+{
+  // capitalize m_detName
+  std::transform(m_detName.begin(), m_detName.end(), m_detName.begin(), ::toupper);
+
+  // default (empty) cellID looper
+  m_loopCellIDs = [] (std::function<void(uint64_t)> lambda) { return; };
+
+  // common objects
+  m_readoutCoder = m_det->readout(m_detName+"Hits").idSpec().decoder();
+  m_detRich      = m_det->detector(m_detName);
+  m_systemID     = m_detRich.id();
+
+  // dRICH readout --------------------------------------------------------------------
+  if(m_detName=="DRICH") {
+
+    // get number of pixels along one side of a SiPM
+    m_num_px = m_det->constant<int>("DRICH_num_px");
+
+    // define cellID looper
+    m_loopCellIDs = [this] (std::function<void(uint64_t)> lambda) {
+      m_log.PrintLog("call ReadoutPixelLoop for systemID = {} = {}", m_systemID, m_detName);
+
+      // loop over sensors (for all sectors)
+      for(auto const& [deName, detSensor] : m_detRich.children()) {
```

Comment:
Excellent idea! @DelloStritto perhaps we make a list of all the possible `cellID`s, then draw _k_ of them for each event. We could use the `ReadoutGeo::VisitAllReadoutPixels` method to make this list; in principle we could get rid of `VisitAllReadoutPixels`, but it may be useful for something else someday, so let's just keep it.

You could make such a list in `AlgorithmInit`, and think about what data structure would be fastest to draw from (e.g., `std::set` would be better than `std::vector`, since order doesn't matter). @wdconinc any suggestions?

Alternatively you could randomly generate the sensor ID number, pixel x, and pixel y, but you'd have to make sure to choose valid numbers. Would it be faster to generate 3 random numbers from small lists vs. 1 random number from a large list?

Reply:
I wouldn't create a list to pick from. Just go with the alternative since in this case your structure is just a uniform cube anyway, without any holes, right? If I recall (from @S-tuberosum's work), there are O(1000^2) pixels, so your list will be O(1M) in memory for a very regular grid. Lookup would be fast, but still it's something to avoid.

Reply:
Yeah it's a 4D cube with no holes. So we have:
- 6 sectors, from DD4hep constant `DRICH_num_sectors`
- almost 1400 sensor modules
  - more than we need, for reasons as discussed in dRICH meetings
  - this number is trickier to get, but you can count them taking inspiration from the `ReadoutGeo::VisitAllReadoutPixels` loop
  - should be the same for each sector, but it would be safe to double check
- 8x8 pixels in x and y, where the number 8 comes from DD4hep constant `DRICH_num_px`
- and finally the fixed dRICH subsystem ID

Once you have those numbers, encode with https://github.com/eic/EICrecon/blob/08b4fa77027eedb2bde1b9404e5f08784de7bc00/src/services/geometry/richgeo/ReadoutGeo.cc#L47-L55
(don't use addition and bitshifting I mentioned earlier, since `x` and `y` are signed (but in practice they only range 0-7))

Reply:
In fact, let's make it easy to get the number of sensors per sector: https://github.com/eic/epic/pull/410
So then you can just use `DRICH_num_sensors` instead of counting.

Reply:
@DelloStritto's latest changes implement this with `RichGeo::VisitAllRngPixels`, but I have some further comments which will appear in new conversations below.